### PR TITLE
Fix the anyhow feature

### LINKED ISSUE #324: `eyre` feature for event handler error handling
Currently there's the `anyhow` feature that allows returning `anyhow::Error`s from event handlers. Implemented here

https://github.com/matrix-org/matrix-rust-sdk/blob/b94faa3121c987a568b531e7d5e5bd033713e9b3/matrix_sdk/src/event_handler.rs#L192-L199

Additionally an `eyre` feature would be helpful to allow returning `eyre::Report`.

Follow-up of https://github.com/matrix-org/matrix-rust-sdk/issues/322#issuecomment-916247273

### LINKED ISSUE #347: `anyhow` feature breaks compilation
**Describe the bug**

```rust
error[E0119]: conflicting implementations of trait `event_handler::EventHandlerResult` for type `std::result::Result<(), anyhow::Error>`
   --> crates/matrix-sdk/src/event_handler.rs:200:1
    |
191 | impl<E: std::error::Error> EventHandlerResult for Result<(), E> {
    | --------------------------------------------------------------- first implementation here
...
200 | impl EventHandlerResult for anyhow::Result<()> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `std::result::Result<(), anyhow::Error>`
    |
    = note: upstream crates may add a new impl of trait `std::error::Error` for type `anyhow::Error` in future versions
```

**To Reproduce**
1. Enable `anyhow` feature
2. Compile

**Expected behavior**
Should compile

### PR DESCRIPTION
**Pull request by jplatte**
Fixes #324
Fixes #347

Supersedes #348

### CODE REVIEW

File: crates/matrix-sdk/src/event_handler.rs [hunk 1]
```diff
@@ -188,19 +192,21 @@ impl EventHandlerResult for () {
     fn print_error(&self, _event_type: &str) {}
 }
 
-impl<E: std::error::Error> EventHandlerResult for Result<(), E> {
+impl<E: fmt::Debug + fmt::Display> EventHandlerResult for Result<(), E> {
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk/src/event_handler.rs:194]: I don't think there's a point in doing this considering that ~everything should be able to have `fmt::Debug` especially if you already have `fmt::Display`.

**Comment by poljar** [comment]
@johannescpk can you confirm if this now works for you?

File: crates/matrix-sdk/src/event_handler.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/matrix-sdk/src/event_handler.rs:194]: Yeah, that's what I was thinking too. Just wanted to point it out that this *could* be relaxed.

**Comment by johannescpk** [comment]
@jplatte Thanks for picking this up. Would you mind to re-export the features in the appservice crate?

Also maybe add the feature to CI in case something breaks again.

**Comment by johannescpk** [comment]
@poljar Yep, looks good to me

**Comment by jplatte** [comment]
I think 18 jobs is already a lot, not sure how to test this in CI without inflating that number / the time CI takes even further.

**Comment by poljar** [comment]
A `cargo check --features=anyhow` fails with this error:
```bash
error[E0310]: the parameter type `E` may not live long enough
   --> crates/matrix-sdk/src/event_handler.rs:198:23
    |
194 | impl<E: fmt::Debug + fmt::Display> EventHandlerResult for Result<(), E> {
    |      -- help: consider adding an explicit lifetime bound...: `E: 'static +`
...
198 |             Err(e) if TypeId::of::<E>() == TypeId::of::<anyhow::Error>() => {
    |                       ^^^^^^^^^^^^^^^ ...so that the type `E` will meet its required lifetime bounds

For more information about this error, try `rustc --explain E0310`.
error: could not compile `matrix-sdk` due to previous error
```
Same thing for the eyre feature:
```bash
error[E0310]: the parameter type `E` may not live long enough
   --> crates/matrix-sdk/src/event_handler.rs:202:23
    |
194 | impl<E: fmt::Debug + fmt::Display> EventHandlerResult for Result<(), E> {
    |      -- help: consider adding an explicit lifetime bound...: `E: 'static +`
...
202 |             Err(e) if TypeId::of::<E>() == TypeId::of::<eyre::Report>() => {
    |                       ^^^^^^^^^^^^^^^ ...so that the type `E` will meet its required lifetime bounds

For more information about this error, try `rustc --explain E0310`.
error: could not compile `matrix-sdk` due to previous error
```

Are we sure that this works?

**Comment by jplatte** [comment]
Ah, damn it I need a proper rust-analyzer config for this repo that enables most / all the features.

**Comment by jplatte** [comment]
This is now technically a breaking change, if that matters.

**Comment by poljar** [comment]
> This is now technically a breaking change, if that matters.

There's a branch for a patch release if we decide to do one, in that case this patch can't be cherry picked to that branch.

**Review by poljar — APPROVED** [review]

**Pull request merged**